lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -28,6 +28,51 @@ using namespace circt;
 using namespace circt::calyx;
 using namespace mlir;
 
+//===----------------------------------------------------------------------===//
+// Utilities related to Direction
+//===----------------------------------------------------------------------===//
+
+Direction direction::get(bool a) { return static_cast<Direction>(a); }
+
+SmallVector<Direction> direction::genInOutDirections(size_t nIns,
+                                                     size_t nOuts) {
+  SmallVector<Direction> dirs;
+  std::generate_n(std::back_inserter(dirs), nIns,
+                  [] { return Direction::Input; });
+  std::generate_n(std::back_inserter(dirs), nOuts,
+                  [] { return Direction::Output; });
+  return dirs;
+}
+
+IntegerAttr direction::packAttribute(ArrayRef<Direction> directions,
+                                     MLIRContext *ctx) {
+  const size_t size = directions.size();
+  // Pack the array of directions into an APInt.  Input is zero, output is one.
+  APInt portDirections(size, 0);
+  for (size_t i = 0, e = directions.size(); i != e; ++i)
+    if (directions[i] == Direction::Output)
+      portDirections.setBit(i);
+
+  return IntegerAttr::get(IntegerType::get(ctx, size), portDirections);
+}
+
+/// Turn a packed representation of port attributes into a vector that can be
+/// worked with.
+SmallVector<Direction> direction::unpackAttribute(Operation *component) {
+  const APInt &value =
+      component->getAttr(direction::attrKey).cast<IntegerAttr>().getValue();
+
+  SmallVector<Direction> result;
+  result.reserve(value.getBitWidth());
```

Comment:
Good use of `reserve`. Perhaps we can save bit width to duplicate calls?

Reply:
Could you elaborate? Not sure what you mean here wrt. duplicate calls.

Reply:
```
auto bitWidth = value.getBitWidth();
result.reserve(bitWidth);
...
for (size_t i = 0, e = bitWidth; i != e; ++i)
  ...
```

Reply:
Ahh, sorry - totally misread your message. I interpreted "save bit width" as in somehow reducing the # of bits stored in the IntegerAttr vector.

Reply:
No problem, I should've been more explicit. Overall LGTM, I'm going to add @mikeurbach for a final review. Thanks again for the contribution @mortbopet ! 

In the mean time, if you're interested in committing this yourself after Mike's approval, please take a look at the [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access).